gadget/install/install.go

```diff
@@ -377,6 +376,140 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
 	}, nil
 }
 
+// structureFromPartDevice returns the OnDiskStructure for a partition
+// node.
+func structureFromPartDevice(diskVol *gadget.OnDiskVolume, partNode string) (*gadget.OnDiskStructure, error) {
+	for _, p := range diskVol.Structure {
+		// We remove /dev from p.Node here
+		if p.Node == partNode {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find partition %q", partNode)
+}
+
+// laidOutStructureForDiskStructure searches for the laid out structure that
+// matches a given OnDiskStructure.
+func laidOutStructureForDiskStructure(laidVols map[string]*gadget.LaidOutVolume, gadgetVolName string, onDiskStruct *gadget.OnDiskStructure) (
+	*gadget.LaidOutStructure, error) {
+
+	for _, laidVol := range laidVols {
+		// Check that this is the right volume
+		if laidVol.Name != gadgetVolName {
+			continue
+		}
+		for _, laidStruct := range laidVol.LaidOutStructure {
+			if onDiskStruct.Name == laidStruct.Name {
+				return &laidStruct, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	syfsLink := filepath.Join("/sys/class/block", filepath.Base(device))
+	partPath, err := os.Readlink(syfsLink)
+	if err != nil {
+		return "", fmt.Errorf("cannot read link %q: %v", syfsLink, err)
+	}
+	// Remove initial ../../ from partPath, and make path absolute
```

Comment:
Is this comment correct? When I look at my system I have:
```
$ ls -l /sys/block/sda
lrwxrwxrwx 1 root root 0 Sep 29 08:56 /sys/block/sda -> ../devices/pci0000:00/0000:00:01.3/0000:02:00.1/ata6/host5/target5:0:0/5:0:0:0/block/sda
```
i.e. only a single `../`.

Reply:
Right, but that is `/sys/block`, not `/sys/class/block`.

Reply:
this code really expects partitions, that will not be true for when we deal with encrpyted partitions otoh we have not content to write to those usually. I think this is ok to unblock us but we might need to rethink this a bit later

Reply:
@mvo is /sys/block/devices another symlink?

Reply:
FTR `/sys/block` does not show partitions, only disks, while `/sys/class/block` does (there is no `/sys/block/devices`)